clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,280 @@
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+			key, err := runtimeobjects.CreateMetaNamespaceKey(&svc)
+			if err != nil {
+				return fmt.Errorf("meta key created error: %v", err)
+			}
+
+			lbSvcMap[key] = svc
+		}
+	}
+	totalLbSvc := len(lbSvcMap)
+
+	// trigger node sync by picking a node and add exclude lb label
+	nodeList, err := s.client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodeList.Items {
+		if !isCandidateNode(node) {
+			continue
+		}
+		s.triggerNodeName = node.Name
+	}
+
+	if s.triggerNodeName == "" {
+		return fmt.Errorf("failed to find a node candidate to trigger nodesync from node list: %v", nodeList.Items)
+	}
+	if err = s.labelNodeForLBs(true); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	for key := range lbSvcMap {
+		s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncStart, now)
+	}
+
+	return wait.Poll(5*time.Second, s.waitTimeout, func() (bool, error) {
+		for key, svc := range lbSvcMap {
+			if _, found := s.svcNodeSyncLatencyTracker.Get(key, phaseNodeSyncCompelte); found {
+				continue
+			}
+
+			eventList, err := s.client.CoreV1().Events(s.selector.Namespace).Search(Scheme, &svc)
+			if err != nil {
+				klog.Errorf("failed to list events associated with svc %v: %v", svc, err)
+				continue
+			}
+
+			for _, event := range eventList.Items {
+				if event.Reason == nodeSyncEventReason {
+					s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncCompelte, event.CreationTimestamp.Time)
+				}
+			}
+		}
+		if s.svcNodeSyncLatencyTracker.Count(phaseNodeSyncCompelte) == totalLbSvc {
+			return true, nil
+		}
+		klog.V(2).Infof("out of a total of %v LBs, %v LB type service has %q event")
+		return false, nil
+	})
+}
+
+// labelNodeForLBs manipulates candidate node to include or exclude it from being LB backends.
+func (s *LoadBalancerNodeSyncMeasurement) labelNodeForLBs(exclude bool) error {
+	ctx := context.Background()
+	node, err := s.client.CoreV1().Nodes().Get(ctx, s.triggerNodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	newNode := node.DeepCopy()
+
+	if exclude {
+		newNode.Labels[labelNodeExcludeBalancers] = "true"
+	} else {
+		delete(newNode.Labels, labelNodeExcludeBalancers)
+	}
+
+	patchBytes, err := preparePatchBytes(node, newNode, v1.Node{})
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.CoreV1().Nodes().Patch(ctx, s.triggerNodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) gather(identifier string) ([]measurement.Summary, error) {
+	klog.V(2).Infof("%s: gathering nodesync latency measurement...", s)
+	nodeSyncLatency := s.svcNodeSyncLatencyTracker.CalculateTransitionsLatency(nodeSyncTransition, measurementutil.MatchAll)
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(nodeSyncLatency))
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Maybe add a TODO to return an error here if the latency is higher than an upper bound (say 30 mins)?

Reply:
Ack

Reply:
I added a waitTimeout in the measurement. The test would fail if it took more than the timeout to measure nodesync.